### PR TITLE
Added GZIP compression support

### PR DESCRIPTION
**Pull request by ankit-splunk**
Added Functionality for GZIP compression.

### CODE REVIEW

File: src/main/java/com/splunk/hecclient/EventBatch.java [hunk 1]
```diff
@@ -42,19 +37,23 @@
 
     private volatile int status = INIT;
     private int failureCount = 0;
+    private boolean enableCompression;
     private long sendTimestamp = System.currentTimeMillis() / 1000; // in seconds
     protected int len;
     protected List<Event> events = new ArrayList<>();
 
     public abstract String getRestEndpoint();
+
```
Comment (chaitanyaphalak) [inline src/main/java/com/splunk/hecclient/EventBatch.java:46]: remove extra lines

File: src/main/java/com/splunk/hecclient/EventBatch.java [hunk 2]
```diff
@@ -42,19 +37,23 @@
...
     public abstract String getContentType();
+
     public abstract void add(Event event);
+
```
Comment (chaitanyaphalak) [inline src/main/java/com/splunk/hecclient/EventBatch.java:50]: remove extra lines

File: src/main/java/com/splunk/hecclient/EventBatch.java [hunk 3]
```diff
@@ -42,19 +37,23 @@
...
     public abstract EventBatch createFromThis();
 
     public final void addExtraFields(final Map<String, String> fields) {
         // recalculate the batch length since we inject more meta data to each event
         int newLength = 0;
-        for (final Event event: events) {
+        for (final Event event : events) {
```
Comment (chaitanyaphalak) [inline src/main/java/com/splunk/hecclient/EventBatch.java:56]: remove extra space

File: src/main/java/com/splunk/hecclient/EventBatch.java [hunk 4]
```diff
@@ -129,18 +128,54 @@ public final HttpEntity getHttpEntity() {
...
+        e.setContentType(getContentType());
+        return e;
+    }
+
     @Override
     public final String toString() {
         StringBuilder builder = new StringBuilder();
         builder.append("[");
-        for (Event e: events) {
+        for (Event e : events) {
```
Comment (chaitanyaphalak) [inline src/main/java/com/splunk/hecclient/EventBatch.java:142]: remove extra space

File: src/test/java/com/splunk/hecclient/IndexerTest.java [hunk 5]
```diff
@@ -194,4 +194,26 @@ private Indexer assertFailure(CloseableHttpClient client) {
         Assert.assertEquals(indexer.getChannel(), poller.getChannel());
         return indexer;
     }
+
+    @Test
+    public void sendCompressedBatchWithSuccess() {
```
Comment (chaitanyaphalak) [inline src/test/java/com/splunk/hecclient/IndexerTest.java:199]: Can we add another test for sending RAW successfully?

File: src/test/java/com/splunk/hecclient/JsonEvenBatchTest.java [hunk 6]
```diff
@@ -177,4 +178,31 @@ private int readContent(final HttpEntity entity, byte[] data) {
 
         return UnitUtil.read(in, data);
     }
+
+    @Test
+    public void testGZIPCompressionForJsonEvent() {
```
Comment (chaitanyaphalak) [inline src/test/java/com/splunk/hecclient/JsonEvenBatchTest.java:183]: Can we add a test case where compression fails?

File: src/test/java/com/splunk/hecclient/RawEventBatchTest.java [hunk 7]
```diff
@@ -122,4 +127,31 @@ public void checkEquals() {
 
         Assert.assertFalse(batchOne.equals(batchTwo));
     }
+
+    @Test
+    public void testGZIPCompressionForRaw() {
```
Comment (chaitanyaphalak) [inline src/test/java/com/splunk/hecclient/RawEventBatchTest.java:132]: Can we add a test case where compression fails?

File: src/main/java/com/splunk/hecclient/EventBatch.java (hunk 1 above)
Comment (ankit-splunk) [reply src/main/java/com/splunk/hecclient/EventBatch.java]: Resolved

File: src/main/java/com/splunk/hecclient/EventBatch.java [hunk 8]
```diff
@@ -42,19 +37,23 @@
...
     private int failureCount = 0;
+    private boolean enableCompression;
     private long sendTimestamp = System.currentTimeMillis() / 1000; // in seconds
     protected int len;
     protected List<Event> events = new ArrayList<>();
 
     public abstract String getRestEndpoint();
+
     public abstract String getContentType();
+
```
[the remark this answers is not in the document]
Comment (ankit-splunk) [reply src/main/java/com/splunk/hecclient/EventBatch.java]: Resolved

File: src/main/java/com/splunk/hecclient/EventBatch.java (hunk 2 above)
Comment (ankit-splunk) [reply src/main/java/com/splunk/hecclient/EventBatch.java]: Resolved

File: src/main/java/com/splunk/hecclient/EventBatch.java (hunk 3 above)
Comment (ankit-splunk) [reply src/main/java/com/splunk/hecclient/EventBatch.java]: Resolved

File: src/main/java/com/splunk/hecclient/EventBatch.java (hunk 4 above)
Comment (ankit-splunk) [reply src/main/java/com/splunk/hecclient/EventBatch.java]: Resolved

File: src/test/java/com/splunk/hecclient/IndexerTest.java (hunk 5 above)
Comment (ankit-splunk) [reply src/test/java/com/splunk/hecclient/IndexerTest.java:199]: Added Test case for Raw batch

Comment (ankit-splunk) [reply src/test/java/com/splunk/hecclient/IndexerTest.java:199]: It's already covered in this test case, so I don’t think to add new test case

File: src/test/java/com/splunk/hecclient/RawEventBatchTest.java (hunk 7 above)
Comment (ankit-splunk) [reply src/test/java/com/splunk/hecclient/RawEventBatchTest.java:132]: It's already covered in this test case, so I don’t think to add a new test case

**Review by chaitanyaphalak — APPROVED** [review]

**Pull request merged**